### PR TITLE
Improve unused includes diagnostic runtime

### PR DESCRIPTION
**Pull request by keynslug**
### Description

By considering only those POIs relevant for this diagnostic, e.g. POIs which MAY potentially be defined in headers. This often significantly cuts number of POIs to iterate over which directly speeds up the diagnostic.

### CODE REVIEW

File: apps/els_lsp/src/els_unused_includes_diagnostics.erl [hunk 1]
```diff
@@ -52,15 +51,24 @@ source() ->
 %%==============================================================================
 %% Internal Functions
 %%==============================================================================
--spec find_unused_includes(els_dt_document:item(), [poi()]) -> [uri()].
-find_unused_includes(#{uri := Uri} = Document, Includes) ->
+-spec find_unused_includes(els_dt_document:item()) -> [uri()].
+find_unused_includes(#{uri := Uri} = Document) ->
   Graph = expand_includes(Document),
-  POIs = els_dt_document:pois(Document),
+  POIs = els_dt_document:pois(Document,
```
Comment (robertoaloi) [inline apps/els_lsp/src/els_unused_includes_diagnostics.erl:57]: I wonder if it's worth maintaining an explicit list of definitions here. Did you measure the gain in scanning only these POIs vs scanning all of them?

**Comment by keynslug** [comment]
Yes, but rather unscientifically. For a sample source file of about 3000
lines long observed speedup was at least 5x on my machine. Nevertheless I'm
more than willing to give you some hard numbers, though no earlier than
next week.

**Comment by robertoaloi** [comment]
Other users reported similar performance issues with these diagnostics, so let's get this in. Thanks!

**Comment by keynslug** [comment]
For the sake of completeness here are some numbers.

### Setup

I've opened a clean workspace with OTP's [`ssl` (as of `OTP-24.0.3`)](https://github.com/erlang/otp/tree/OTP-24.0.3/lib/ssl) application in VS Code with erlang-ls enabled, first time with this PR applied, second time with vanilla 0.17.0.

Both times I ran the following from a remote shell:
```
Uris = [ % source files with most includes
    <<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/dtls_gen_connection.erl">>,
    <<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/tls_connection.erl">>,
    <<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/tls_handshake.erl">>
],
_ = els_unused_includes_diagnostics:run(hd(Uris)), % lazy warmup
[{Uri, element(1, timer:tc(lists, foreach, [fun (_) -> els_unused_includes_diagnostics:run(Uri) end, lists:seq(1, 10)]))} || Uri <- Uris].
```

Raw results under this PR:
```
[{<<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/dtls_gen_connection.erl">>, 5304794},
 {<<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/tls_connection.erl">>, 4706666},
 {<<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/tls_handshake.erl">>, 2771074}]
```

Raw results under 0.17.0:
```
[{<<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/dtls_gen_connection.erl">>, 24562051},
 {<<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/tls_connection.erl">>, 15740154},
 {<<"file:///home/keynslug/ws/_3p/erlang/lib/ssl/src/tls_handshake.erl">>, 10424297}]
```

### Summary

| module              | avg PR (µs) | avg 0.17.0 (µs) | speedup |
| ------------------- | ----------- | --------------- | ------- |
| dtls_gen_connection | 530479.4    | 2456205.1       | 4.63x   |
| tls_connection      | 470666.6    | 1574015.4       | 3.34x   |
| tls_handshake       | 277107.4    | 1042429.7       | 3.76x   |

### Environment

Hardware: i7-8550U CPU @ 1.80GHz w/ Turbo Boost off
OS: Linux 5.12.15-arch1-1
Erlang: Erlang/OTP 24 [erts-12.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

**Comment by robertoaloi** [comment]
Thanks so much for the follow-up @keynslug !

**Pull request merged**